File: solutions.digamma.damas.jcr/src/main/java/solutions/digamma/damas/jcr/repo/BuiltInJob.java
package solutions.digamma.damas.jcr.repo;

import solutions.digamma.damas.jcr.names.TypeNamespace;
import solutions.digamma.damas.jcr.content.JcrFile;

import java.util.Arrays;
import java.util.List;

/**
 * Built-in repository job.
 *
 * @author Ahmad Shahwan
 */
public final class BuiltInJob implements RepositoryJob {

    public static final BuiltInJob INSTANCE = new BuiltInJob();

    private final static Node[] CREATIONS = {
            new Node(
                    JcrFile.ROOT_PATH,
                    TypeNamespace.FOLDER,
                    new String[0]
            )
    };

    /**
     * Constructor.
     */
    private BuiltInJob() {
    }

    /**
     * List convert nodes to be created.
     *
     * @return
     */
    public List<RepositoryJob.Node> getCreations() {
        return Arrays.asList(CREATIONS);
    }

    /**
     * Node specification.
     */
    private static class Node implements RepositoryJob.Node {

        private final String path;
        private final String type;
        private final String[] mixins;

        public Node(String path, String type, String[] mixins) {
            this.path = path;
            this.type = type;
            this.mixins = mixins;
        }

        @Override
        public String getPath() {
            return this.path;
        }

        @Override
        public String getType() {
            return this.type;
        }

        @Override
        public List<String> getMixins() {
            return Arrays.asList(this.mixins);
        }
    }
}

File: solutions.digamma.damas.jcr/src/main/kotlin/solutions/digamma/damas/jcr/repo/BuiltInJob.kt
package solutions.digamma.damas.jcr.repo

import solutions.digamma.damas.jcr.content.JcrFile
import solutions.digamma.damas.jcr.names.TypeNamespace
import solutions.digamma.damas.jcr.user.JcrSubject
import java.util.Collections

/**
 * Built-in repository job.
 *
 * @author Ahmad Shahwan
 */
internal object BuiltInJob : RepositoryJob {

    /**
     * Node specification.
     */
    private class Node(
            override val path: String,
            override val type: String,
            override val mixins: List<String> = Collections.emptyList()
    ): RepositoryJob.Node

    /**
     * List convert nodes to be created.
     *
     * @return
     */
    override val creations: List<RepositoryJob.Node> = listOf(
        Node(JcrFile.ROOT_PATH, TypeNamespace.FOLDER),
        Node(JcrSubject.ROOT_PATH, TypeNamespace.DIRECTORY)
    )
}
